src/main/java/org/testng/internal/Parameters.java

```diff
@@ -264,35 +265,81 @@ else if (ITestResult.class.equals(c)) {
                                        MethodParameters params,
                                        XmlSuite xmlSuite) {
     List<Object> vResult = Lists.newArrayList();
-    for (int i = 0, j = 0; i < parameterTypes.length; i++) {
-      if (j < parameterNames.length) {
-        String p = parameterNames[j];
-        String value = params.xmlParameters.get(p);
-        if (null == value) {
-          // try SysEnv entries
-          value = System.getProperty(p);
+    if (isInjected) {
+      FilterOutInJectedTypesResult filterOutResult = FilterOutInJectedTypesFromOptionalValues(parameterTypes, optionalValues);
+      optionalValues = filterOutResult.getOptionalValues();
+      parameterTypes = filterOutResult.getParameterTypes();
+    }
+    for (int i = 0; i < parameterNames.length; i++) {
+      String p = parameterNames[i];
+      String value = params.xmlParameters.get(p);
+      if (null == value) {
+        // try SysEnv entries
+        value = System.getProperty(p);
+      }
+      if (null == value) {
+        if (optionalValues != null) {
+            value = optionalValues[i];
         }
         if (null == value) {
-          if (optionalValues != null) {
-            value = optionalValues[i];
-          }
-          if (null == value) {
             throw new TestNGException("Parameter '" + p + "' is required by "
                     + methodAnnotation
                     + " on " + prefix + " "
                     + name
                     + " but has not been marked @Optional or defined\n"
                     + (xmlSuite.getFileName() != null ? "in "
-                    + xmlSuite.getFileName() : ""));
-          }
+                            + xmlSuite.getFileName() : ""));
+        }
+    }
+      vResult.add(convertType(parameterTypes[i], value, p));
+    }
+    return vResult;
+  }
+  
+  /**
+   * Remove injected types from parameterTypes and optionalValues
+   * @param parameterTypes
+   * @param optionalValues
+   * @return FilterOutInJectedTypesResult
+   */
+  private static FilterOutInJectedTypesResult FilterOutInJectedTypesFromOptionalValues(Class<?>[] parameterTypes, String[] optionalValues) {
+    List<Class<?>> typeList = Lists.newArrayList();
+    List<String> optionalValueList = Lists.newArrayList();
+    Collections.addAll(typeList, parameterTypes);
+    Collections.addAll(optionalValueList, optionalValues);
+    Iterator<Class<?>> typeIterator = typeList.iterator();
+    Iterator<String> optionalIterator = optionalValueList.iterator();
+    while (typeIterator.hasNext()) {
+        Class<?> parameterType = typeIterator.next();
+        optionalIterator.next();
+        if (INJECTED_TYPES.contains(parameterType)) {
+            optionalIterator.remove();
+            typeIterator.remove();
         }
+    }
+    return new FilterOutInJectedTypesResult(typeList.toArray(new Class<?>[0]),
```

Comment:
Do we really need to be doing this conversion from `List` to `Array` especially when all we are going to be doing is retrieval via indexes, which is supported in `List` anyways ?

Reply:
I am thinking about it. But if we return back as List, it means that `optionalValues `and `parameterNames `from `Parameters::createParam` has to be converted to List first. That is fine if had injected parameters in test case. But if test case doesn't have injected parameters, such covertion would be a waste.